### DISCUSSION TITLE
Authentication using API Key and Secret

### DISCUSSION THREAD

**Discussion by Bunnywood** [CoreTweet/CoreTweet#195] [Q&A] [answered]
Hi,
Nice toolset. Is there a way to authenticate console applications, basically key and secret, or bearer tokens, without the PIN number?

Thanks.
Rob.

**Comment by azyobuzin** [comment]
Do you want to use [application-only authentication](https://developer.twitter.com/en/docs/authentication/oauth-2-0/application-only)? CoreTweet supports it by [OAuth2 class](http://coretweet.github.io/docs/classCoreTweet_1_1OAuth2.html).

**Reply by Bunnywood** [reply]
Hi, Thanks for the very quick reply!

Yes, thats exactly what I want I believe.

I've tried this (VB, sorry!):
            Dim Token2 = OAuth2.GetToken(APIKey, APIKeySecret)
            Token2.Statuses.Update(New With {.status = "Testing ... "})

The authentication seems to go through, but I keep getting forbidden when it trys to tweet, even though I have double and triple checked the values from the dev portal.

I also tried this with similar results:
            Dim Token2 = CoreTweet.OAuth2Token.Create(APIKey, APIKeySecret, BearerToken)

Now, having gone through the twitter doc's again, I wonder if I need a second level of authentication to send the tweet?

> Please note that OAuth 1.0a is still required to issues requests on behalf of users. The API reference page describes the authentication method required to use an API. You will need user-authentication, user-context, with an access token to perform the following:
> 
> Post Tweets or other resources
> Search for users
> Use any geo endpoint
> Access Direct Messages or account credentials
> Retrieve user's email addresses

This page seems to suggest I need a token and token secret too, which I have, but have no idea how to add them:
https://developer.twitter.com/en/docs/authentication/oauth-1-0a

Any ideas?

Thanks

**Comment by cannorin** [comment, marked as answer]
There is two ways to get an OAuth1.0a token: PIN code and callback.

## PIN code

I think **it would be a lot easier to use PIN codes** since you are building a console application.

See: https://github.com/CoreTweet/CoreTweet/wiki/Authentication

Note that you would need to do it only once: https://github.com/CoreTweet/CoreTweet/wiki/Saving-and-Reusing-Tokens-Howto

## Callback

If using PIN code isn't an option for you, **you would need to have a (temporary) local web server** to receive a callback request from Twitter. So **I don't recommend using this if you are not familiar with web programming in .NET**.

See the Step 2 of this official docs: https://developer.twitter.com/en/docs/authentication/guides/log-in-with-twitter

Note that the rest of the authentication flow is the same as the PIN-code flow.
The only difference is you will receive your `oauth_verifier` (equivalent to PIN code) via the GET request from Twitter.

```csharp
tokens = session.GetTokens(oauth_verifier);
```

**Reply by Bunnywood** [reply]
Hi,

Thanks for that, had not found that coretweet documentation page somehow, and its really helpful.

I now have a PIN number!

Still not working completely though as I get this as an error now when calling GetTokens:

EXCEPTION: Error processing your OAuth request: Invalid oauth_verifier parameter
CoreTweet.TwitterException: Error processing your OAuth request: Invalid oauth_verifier parameter
 ---> System.Net.WebException: The remote server returned an error: (401) Unauthorized.
   at System.Net.HttpWebRequest.GetResponse()
   at CoreTweet.Request.HttpPost(Uri url, String contentType, Byte[] content, String authorizationHeader, ConnectionOptions options)
   at CoreTweet.Request.HttpPost(Uri url, IEnumerable`1 prm, String authorizationHeader, ConnectionOptions options)
   at CoreTweet.OAuth.GetTokens(OAuthSession session, String oauthVerifier)

```
    Try
            ' https://github.com/CoreTweet/CoreTweet/wiki/Authentication
            Console.WriteLine("Authorisation ... ")
            Dim Session = OAuth.Authorize(APIKey, APIKeySecret)

            If APIPinCode = "" Then
                Console.WriteLine(Session.AuthorizeUri.AbsoluteUri)
                System.Diagnostics.Process.Start("Chrome.exe", Session.AuthorizeUri.AbsoluteUri)

                Console.WriteLine("Capture the PIN number from the web page now shown. You will need to add it to the code/config.")
                Console.WriteLine("Press any key once captured and stored.")
                Console.ReadKey()

            Else
                Console.WriteLine("Tokens ... ")
                Dim Tokens = OAuth.GetTokens(Session, APIPinCode)       '  <----- ERROR HERE

                Console.WriteLine("Tweeting ... ")
                Tokens.Statuses.Update(New With {.status = "Testing ... "})

                Console.WriteLine("Done!")

            End If


            'Dim Token2 = CoreTweet.OAuth2Token.Create(APIKey, APIKeySecret, BearerToken)
            'Dim Token2 = OAuth2.GetToken(APIKey, APIKeySecret)
            'Dim Token2 = OAuth2.GetToken(OAuthToken, OAuthTokenSecret)

        Catch ex As Exception
            Console.WriteLine(String.Format("EXCEPTION: {0}", ex.Message))
            Console.WriteLine(ex.ToString)

        End Try

```

Thanks for the help so far :)

**Reply by Bunnywood** [reply]
OK, I see my mistake. It's all inline, get the PIN first, then feed it into GetToken. Can reuse PIN's.

Thanks for your help, much appreciated!

**Reply by cannorin** [reply]
I'm glad it helped, just note that you can't reuse the PIN code itself but you can reuse the `Tokens.AccessToken` and `Tokens.AccessTokenSecret` and then simply do `Dim Tokens = Tokens.Create(APIKey, APIKeySecret, AccessToken, AccessTokenSecret)` without any authentication.
